### PR TITLE
PR: Enable LSP autoformatting support

### LINKED ISSUE #11396: Integration of automatic formatting tools: autopep8, yapf and black
With the Python extension of Visual Studio Code, it is by default possible to format a file just by pressing keys. The same feature should be available in Spyder.

It is especially useful for students and new learners that just discover the PEP 8. But it is actually very useful for all Python developers...

This issue is similar to #9359 (which was closed after a very strange discussion about the dark theme, which has nothing to do with code formatting).

There are plugins (https://github.com/spyder-ide/spyder-autopep8 and https://github.com/kikocorreoso/spyder-black-formatter) but they are broken with Spyder 4.

In particular, they contain `from spyder.config.gui import fixed_shortcut`, which does not work anymore. `fixed_shortcut` was depreciated in https://github.com/spyder-ide/spyder/commit/1b09f24fe02bc955693e8cab38790d6da5bbced0 (but only with a warning in the docstring).

Anyway, this feature is so important (and not difficult to implement with autopep8 and black) that it could be available by default.

### PR DESCRIPTION
**Pull request by andfoy**
## Description of Changes
This PR introduces the option to autoformat files/selection regions via the LSP `textDocument/formatting` and `textDocument/rangeFormatting` calls.

## Pending tasks
- [x]  Implement `textDocument/formatting`
- [x]  Implement `textDocument/rangeFormatting`
- [x]  Add option to contextual menus and toolbar
- [x]  Add shortcut 
- [x]  Add pyls-black, yapf and autopep8 to dependencies
- [x]  Add tests

## Contribution checklist
* [x] Wrote at least one-line docstrings (for any new functions)
* [x] Added unit test(s) covering the changes (if testable)

* [x] Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/))

#### Editor context menu

imagen

#### Source menu option

imagen

#### Editor configuration page options to remove/add trailling newlines

imagen
 
#### Code formatting tab under completion and linting configuration page

imagen

#### Format whole document (autopep8)

Peek 17-09-2020 11-38

#### Format selection (autopep8)

Peek 17-09-2020 11-42

#### Format whole document (yapf)

Peek 17-09-2020 11-44

#### Format selection (yapf)

Peek 17-09-2020 11-47

#### Format whole document (black)

Peek 17-09-2020 11-51

#### Format selection (black)

Peek 17-09-2020 11-53

#### Format document with partial document updates using three-merge (Elixir)

Peek 17-09-2020 12-05

### Issue(s) Resolved

Fixes #11396.

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct: @andfoy

### CODE REVIEW

**Review by ccordoba12 — COMMENTED** [review]
Thanks @andfoy for this! It looks pretty good!

Other than my current comments, I have some important remarks:

* After selecting the formatter provider in our Preferences, we should also set the max line length in the Pycodestyle section (88 for Black, 79 for Autopep8 and Yapf)
* Also after changing the formatter, we should set the vertical line in the editor that signals the max line length with those values.
* Finally, if the user selects a different line length in our Preferences (either in the Editor or the Pycodestyle section), we should update the formatter config options with that value. Autopep8 and Pyls-black can already be configured through the PyLS, so we'd need to add that functiionality to the Yapf too.

File: spyder/plugins/completion/languageserver/providers/document.py [hunk 1]
```diff
@@ -288,3 +288,49 @@ def document_did_close(self, params):
                 self.watched_files.pop(filename)
 
         return params
+
+    @send_request(method=LSPRequestTypes.DOCUMENT_FORMATTING)
+    def document_formatting_request(self, params):
+        options = params['options']
+        options = {
+            snake_to_camel(opt): options[opt]
```
[the remark this answers is not in the document]
Comment (andfoy) [reply spyder/plugins/completion/languageserver/providers/document.py:296]: The LSP spec defines those parameters in CamelCase, and using CamelCase in Python is strange

File: spyder/plugins/editor/widgets/tests/test_formatting.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+@pytest.mark.first
+@pytest.mark.parametrize('formatter', ['autopep8', 'yapf', 'black'])
+def test_document_formatting(formatter, lsp_codeeditor, qtbot):
+    """Validate text autoformatting via autopep8, yapf or black."""
+    code_editor, manager = lsp_codeeditor
+    text, expected = get_formatter_values(formatter)
+
+    # After this call the manager needs to be reinitialized
+    CONF.set('lsp-server', 'formatting', formatter)
+    qtbot.wait(2000)
```
[the remark this answers is not in the document]
Comment (andfoy) [reply spyder/plugins/editor/widgets/tests/test_formatting.py:50]: Update configuration takes the value set in the configuration, this is a common pattern found on other LSP tests such as https://github.com/spyder-ide/spyder/blob/8923aa9c0a47665eb5f6111f14a021d625899ed2/spyder/plugins/editor/widgets/tests/test_introspection.py#L601

[the remark this answers is not in the document]
Comment (ccordoba12) [reply spyder/plugins/editor/widgets/tests/test_formatting.py:50]: Ok, but I think setting a new value in our config system doesn't trigger an update of PyLS configuration.

However, this is not a big problem, so let's leave things like this.

**Comment by andfoy** [comment]
Complementary PR in pyls: https://github.com/palantir/python-language-server/pull/866

**Review by ccordoba12 — COMMENTED** [review]
Thanks @andfoy for the fixes and improvements. This is my second round of comments.

Besides the minor style issues I mentioned before, I have the following comments about the UI elements introduced in this PR:

* The new entries in `Source` and the editor's context menu turned out to be too long (my bad, sorry). Please shorten them to `Format file or selection with ...`. Also, please capitalize the name of the formatter (right now is not):

    Selección_001

* The text in those entries is not updated when you select a different formatter. So, if you select Black, the name should be `Format file or selection with Black`. For the `Source` menu you probably need to update it using the `AboutToShow` signal. I don't know how to do the same for the context menu.

* After the file is autoformatted, the name on its corresponding tab ends up being surrounded by parenthesis. This is the way that we signal a file is read-only, so that's wrong. Instead, please make the tab's name appear surrounded by parenthesis while the spinner is running and remove them afterwards.

File: spyder/plugins/editor/plugin.py [hunk 3]
```diff
@@ -811,8 +811,8 @@ def get_plugin_actions(self):
         formatter = CONF.get('lsp-server', 'formatting')
         self.formatting_action = create_action(
             self,
-            _("Format the current file or selection with {0}").format(
-                formatter),
+            _('Format the current file or selection with {0}').format(
```
Comment (ccordoba12) [inline spyder/plugins/editor/plugin.py:814]: Please change this to

    Format file or selection with {0}

File: spyder/plugins/editor/plugin.py [hunk 4]
```diff
@@ -1656,6 +1656,12 @@ def refresh_eol_chars(self, os_name):
     def refresh_formatting(self, status):
         self.formatting_action.setEnabled(status)
 
+    def refresh_formatter_name(self):
+        formatter = CONF.get('lsp-server', 'formatting')
+        self.formatting_action.setText(
+            _('Format the current file or selection with {0}').format(
```
Comment (ccordoba12) [inline spyder/plugins/editor/plugin.py:1662]: Please change this to

    Format file or selection with {0}

File: spyder/plugins/editor/widgets/codeeditor.py [hunk 5]
```diff
@@ -4040,8 +4040,9 @@ def setup_context_menu(self):
         # Document formatting
         formatter = CONF.get('lsp-server', 'formatting')
         self.format_action = create_action(
-            self, _('Format the current file or selection with {0}').format(
-                formatter),
+            self,
+            _('Format the current file or selection with {0}').format(
```
Comment (ccordoba12) [inline spyder/plugins/editor/widgets/codeeditor.py:4065]: Please change this to

    Format file or selection with {0}

File: spyder/plugins/editor/widgets/codeeditor.py [hunk 6]
```diff
@@ -4817,6 +4818,11 @@ def contextMenuEvent(self, event):
         self.re_run_last_cell_action.setVisible(self.is_python())
         self.gotodef_action.setVisible(self.go_to_definition_enabled)
 
+        formatter = CONF.get('lsp-server', 'formatting')
+        self.format_action.setText(_(
+            'Format the current file or selection with {0}').format(
```
Comment (ccordoba12) [inline spyder/plugins/editor/widgets/codeeditor.py:4844]: Please change this to

    Format file or selection with {0}

**Review by ccordoba12 — COMMENTED** [review]
Thanks for the latest changes @andfoy! I think this is almost ready!

Besides my comments below, I found two important bugs.

* If the file being formatted doesn't have a blank line at the end, like in the image below

    Selección_007

    then after the formatting request finishes, we end up with the formatted and original texts in the same file

    Selección_009

    (You can see that the original text is exactly below where the formatted one ends).

* If the original file ends with a blank line

    Selección_010

    we are inserting an extra blank line after the formatting finishes

    Selección_011

**Review by ccordoba12 — APPROVED** [review]
Thanks a lot @andfoy for your hard work on this one! What a great addition!

**Comment by ccordoba12** [comment]
@spyder-ide/core-developers, @goanpeca, this PR adds two new dependencies to Spyder: `pyls-black` and `three-merge`. You can install them by running:

    conda install -c spyder-ide/label/dev pyls-black three-merge

**Pull request merged**